src/validation/variable.rs

```diff
@@ -1,16 +1,123 @@
-use plc_ast::ast::{ArgumentProperty, Pou, PouType, Variable, VariableBlock, VariableBlockType};
+use plc_ast::ast::{
+    ArgumentProperty, ConfigVariable, Pou, PouType, Variable, VariableBlock, VariableBlockType,
+};
 use plc_diagnostics::diagnostics::Diagnostic;
 
 use super::{
     array::validate_array_assignment,
-    statement::{validate_enum_variant_assignment, validate_pointer_assignment, visit_statement},
+    statement::visit_statement,
     types::{data_type_is_fb_or_class_instance, visit_data_type_declaration},
     ValidationContext, Validator, Validators,
 };
 use crate::index::VariableIndexEntry;
 use crate::typesystem::DataTypeInformation;
+use crate::validation::statement::validate_enum_variant_assignment;
+use crate::validation::statement::validate_pointer_assignment;
 use crate::{index::const_expressions::ConstExpression, resolver::AnnotationMap};
 
+pub fn visit_config_variable<T: AnnotationMap>(
+    validator: &mut Validator,
+    var_config: &ConfigVariable,
+    context: &ValidationContext<T>,
+) {
+    if var_config.name_segments.is_empty() {
+        return;
+    }
+
+    let segments: Vec<&str> = var_config.name_segments.iter().map(|segment| segment.as_str()).collect();
+    let segments: &[&str] = &segments;
+
+    match context.index.find_variable(Some(segments[0]), &segments[1..]) {
+        // The template variable referenced in the VAR_CONFIG block does not exist
+        None => {
+            validator.push_diagnostic(
+                Diagnostic::new(format!(
+                    "Template variable `{}` does not exist",
+                    segments.last().expect("must exist due to previous length check")
+                ))
+                .with_error_code("E101")
+                .with_location(&var_config.location),
+            );
+        }
+
+        // The template variable does exist, check
+        // (1) if the types of the config and template variable are the same
+        // (2) if the template variable has a hardware binding (`.. AT ... : ...`)
+        // (3) if the config variable has specified a full hardware address
+        // (4) if the template variable has specified a incomplete hardware address
+        Some(var_template) => {
+            // (1)
+            let (var_config_ty, var_config_ty_info) = {
+                let ty_name = &var_config.data_type.get_name().unwrap_or_default();
+                let ty = context.index.get_effective_type_or_void_by_name(ty_name);
+                let ty_info = ty.get_type_information();
+
+                (ty, ty_info)
+            };
+
+            let (var_template_ty, var_template_ty_info) = {
+                let ty = context.index.get_effective_type_or_void_by_name(&var_template.data_type_name);
+                let ty_info = context.index.find_elementary_pointer_type(ty.get_type_information());
+
+                (ty, ty_info)
+            };
+
+            if var_template_ty_info != var_config_ty_info {
```

Comment:
Probably not really an issue, but when declaring `STRING`s with different lengths, e.g.
```
FUNCTION_BLOCK foo_fb
    VAR
        bar AT %I* : STRING[100];
    END_VAR
END_FUNCTION_BLOCK

PROGRAM main
    VAR
        foo : foo_fb;
    END_VAR
END_PROGRAM

VAR_CONFIG
    main.foo.bar AT %IX1.0 : STRING;
END_VAR
```
this will lead to an incompatible types error. Since these are hardware addresses, I don't think strings should be too much of an issue here. Technically, the types are different. But we don't give out errors for different string types elsewhere, so we might not want to here either for consistency sake.

Reply:
Do we want to be extra strict in the context of hardware addresses and just allow certain datatypes (integers, booleans), yielding errors for everything else?

Reply:
I like the idea, just need to double-check what data we can expect via the hardware channels.